From 2fcbc820e6e37d3267cfe95350688a519a2fac2e Mon Sep 17 00:00:00 2001
From: Alon Zakai <azakai@google.com>
Date: Wed, 15 Mar 2023 11:52:35 -0700
Subject: [PATCH] Fix for-of in acorn optimizer (#18968)

The JS AST has a separate node type for for-of, like for-in.

Fixes #18967
---
 test/optimizer/JSDCE-fors-output.js | 2 ++
 test/optimizer/JSDCE-fors.js        | 1 +
 tools/acorn-optimizer.js            | 3 +++
 3 files changed, 6 insertions(+)

diff --git a/test/optimizer/JSDCE-fors-output.js b/test/optimizer/JSDCE-fors-output.js
index ec992dd36..45e46242c 100644
--- a/test/optimizer/JSDCE-fors-output.js
+++ b/test/optimizer/JSDCE-fors-output.js
@@ -1,3 +1,5 @@
 for (var i in x) {}
 
+for (var i of [ 1, 2, 3 ]) {}
+
 for (var j = 0; ;) {}
diff --git a/test/optimizer/JSDCE-fors.js b/test/optimizer/JSDCE-fors.js
index 71e801025..642fc9358 100644
--- a/test/optimizer/JSDCE-fors.js
+++ b/test/optimizer/JSDCE-fors.js
@@ -1,4 +1,5 @@
 
 for (var i in x) {}
+for (var i of [1, 2, 3]) {}
 for (var j = 0;;) {}
 
diff --git a/tools/acorn-optimizer.js b/tools/acorn-optimizer.js
index a44cfc06d..a6874c04c 100755
--- a/tools/acorn-optimizer.js
+++ b/tools/acorn-optimizer.js
@@ -209,6 +209,9 @@ function restoreForVars(node) {
     ForInStatement(node) {
       fix(node.left);
     },
+    ForOfStatement(node) {
+      fix(node.left);
+    },
   });
   return restored;
 }
-- 
2.34.1

